### PR TITLE
feat: ⌨️ Popconfirm can be closed by pressing ESC

### LINKED ISSUE #15203: Closing Popconfirm using esc key
- [x] I have searched the [issues](https://github.com/ant-design/ant-design/issues) of this repository and believe that this is not a duplicate.

### What problem does this feature solve?
When popconfirm is opened, the only way to close it is by action buttons (ok/cancel). I wish the user could also press esc key which will have the same behavior as 'cancel' button.

### What does the proposed API look like?
'closable' boolean prop

### PR DESCRIPTION
**Pull request by afc163**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Perfermance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 💡 Background and solution

close #15203
### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | Popconfirm can be closed by pressing `ESC` now. |
| 🇨🇳 Chinese | Popconfirm 支持按 `ESC` 关闭。 |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/popconfirm/index.tsx [hunk 1]
```diff
@@ -72,6 +91,12 @@ const Popconfirm = React.forwardRef<unknown, PopconfirmProps>((props, ref) => {
     }
   };
 
+  const onKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
+    if (e.key === 'Escape' && e.target === innerContentRef.current) {
+      settingVisible(false, e);
+    }
```
Comment (zpr1g) [inline components/popconfirm/index.tsx:97]: 还要判断一下 `e.target === innerContentRef.current` ，学习了 😂

File: components/popconfirm/index.tsx [hunk 2]
```diff
@@ -46,7 +49,23 @@ const Popconfirm = React.forwardRef<unknown, PopconfirmProps>((props, ref) => {
...
+  React.useEffect(() => {
+    const timeout = setTimeout(() => {
+      if (visible) {
+        innerContentRef.current?.focus();
+      }
+    });
+    return () => {
+      clearTimeout(timeout);
+    };
+  }, [visible]);
```
Comment (zpr1g) [inline components/popconfirm/index.tsx:63]: 如果在 Popconfirm 受控的情况下，切换了 focus 就不能按 Esc 退出了吧？

找了下默认是怎么做到在外部点击关闭弹窗的，应该是[这里](https://github.com/react-component/trigger/blob/master/src/index.tsx#L384) 。那么即使在受控的情况下，也不用自己监听外部点击事件了，配置 `onVisibleChange` 就行了啊。之前还琢磨怎么自己监听实现呢 \_(:3J∠)\_

**Review by yunsii — APPROVED** [review]

File: components/popconfirm/index.tsx [hunk 3]
```diff
@@ -46,7 +49,23 @@ const Popconfirm = React.forwardRef<unknown, PopconfirmProps>((props, ref) => {
     }
   }, [props.defaultVisible]);
 
-  const settingVisible = (value: boolean, e?: React.MouseEvent<HTMLButtonElement>) => {
+  const innerContentRef = React.useRef<HTMLDivElement>(null);
+
+  React.useEffect(() => {
+    const timeout = setTimeout(() => {
+      if (visible) {
+        innerContentRef.current?.focus();
```
Comment (zombieJ) [inline components/popconfirm/index.tsx:57]: 这样键盘交互焦点就丢了，是不是可以考虑全局监听一下。打开的时候 ESC 关闭？

Comment (afc163) [reply components/popconfirm/index.tsx:57]: 直接绑定在 trigger 上好了，全局监听就复杂了。

File: components/popconfirm/index.tsx [hunk 4]
```diff
@@ -72,6 +78,12 @@ const Popconfirm = React.forwardRef<unknown, PopconfirmProps>((props, ref) => {
     }
   };
 
+  const onKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
+    if (e.key === 'Escape' && visible) {
```
Comment (zombieJ) [inline components/popconfirm/index.tsx:82]: KeyCode.ESC

File: components/popconfirm/index.tsx [hunk 5]
```diff
@@ -125,7 +137,16 @@ const Popconfirm = React.forwardRef<unknown, PopconfirmProps>((props, ref) => {
       visible={visible}
       overlay={overlay}
       ref={ref as any}
-    />
+    >
+      {React.isValidElement(children)
```
Comment (zombieJ) [inline components/popconfirm/index.tsx:141]: 用 `components/_util/reactNode` 的 `replaceElement` 减少代码量。

Comment (afc163) [reply components/popconfirm/index.tsx]: 改成 `components/_util/reactNode` 里的 cloneElement 了

**Comment by afc163** [comment]
超了 0.1 kB..

**Comment by afc163** [comment]
Need  to approve

**Review by yoyo837 — APPROVED** [review]

**Pull request merged**